core/src/ldml/ldml_transforms.cpp

```diff
@@ -7,15 +7,262 @@
 
 #include "ldml_transforms.hpp"
 #include "debuglog.h"
+#include <algorithm>
+#include <string>
 
 #ifndef assert
-#define assert(x) // TODO-LDML
+#define assert(x)  // TODO-LDML
 #endif
 
 namespace km {
 namespace kbp {
 namespace ldml {
 
+element::element(const USet &new_u, KMX_DWORD new_flags)
+    : chr(), uset(new_u), flags((new_flags & ~LDML_ELEM_FLAGS_TYPE) | LDML_ELEM_FLAGS_TYPE_USET) {
+}
+
+element::element(km_kbp_usv ch, KMX_DWORD new_flags)
+    : chr(ch), uset(), flags((new_flags & ~LDML_ELEM_FLAGS_TYPE) | LDML_ELEM_FLAGS_TYPE_CHAR) {
+}
+
+bool
+element::is_uset() const {
+  return (flags & LDML_ELEM_FLAGS_TYPE) == LDML_ELEM_FLAGS_TYPE_USET;
+}
+
+signed char
+element::get_order() const {
+  unsigned char uorder = ((flags & LDML_ELEM_FLAGS_ORDER_MASK) >> LDML_ELEM_FLAGS_ORDER_BITSHIFT);
+  return (signed char)uorder;
+}
+
+signed char
+element::get_tertiary() const {
+  unsigned char uorder = ((flags & LDML_ELEM_FLAGS_TERTIARY_MASK) >> LDML_ELEM_FLAGS_TERTIARY_BITSHIFT);
+  return (signed char)uorder;
+}
+
+bool
+element::is_prebase() const {
+  return flags & LDML_ELEM_FLAGS_PREBASE;
+}
+
+bool
+element::is_tertiary_base() const {
+  return flags & LDML_ELEM_FLAGS_TERTIARY_BASE;
+}
+
+KMX_DWORD
+element::get_flags() const {
+  return flags;
+}
+
+bool
+element::matches(km_kbp_usv ch) const {
+  if (is_uset()) {
+    return uset.contains(ch);
+  } else {
+    return chr == ch;
+  }
+}
+
+int
+reorder_sort_key::compare(const reorder_sort_key &other) const {
+  if (primary < other.primary) {
+    return -1;
+  } else if (primary > other.primary) {
+    return 1;
+  } else if (secondary < other.secondary) {
+    return -1;
+  } else if (secondary > other.secondary) {
+    return 1;
+  } else if (tertiary < other.tertiary) {
+    return -1;
+  } else if (tertiary > other.tertiary) {
+    return 1;
+  } else if (quaternary < other.quaternary) {
+    return -1;
+  } else if (quaternary > other.quaternary) {
+    return 1;
+  } else if (ch < other.ch) {  // tiebreak with char value
+    return -1;
+  } else if (ch > other.ch) {
+    return 1;
+  } else {
+    return 0;  // identical
+  }
+}
+
+bool
+reorder_sort_key::operator<(const reorder_sort_key &other) const {
+  return (compare(other) == -1);
```

Comment:
nit ahead!
While we know that `compare` returns -1, 0, or 1, typically a `compare()` contract is only `<0`, `0`, or `>0` so I wonder if we should be:

```suggestion
  return (compare(other) < 0);
```